src/commands/publish/preview/mod.rs

```diff
@@ -10,63 +10,158 @@ use uuid::Uuid;
 
 use crate::commands;
 use crate::http;
+use crate::settings::global_user::GlobalUser;
 use crate::settings::project::Project;
 use crate::terminal::message;
+use reqwest::Client;
+
+// Using this instead of just `https://cloudflareworkers.com` returns just the worker response to the CLI
+const PREVIEW_ADDRESS: &str = "https://00000000000000000000000000000000.cloudflareworkers.com";
 
 #[derive(Debug, Deserialize)]
 struct Preview {
     pub id: String,
 }
 
+impl From<ApiPreview> for Preview {
+    fn from(api_preview: ApiPreview) -> Preview {
+        Preview {
+            id: api_preview.preview_id,
+        }
+    }
+}
+
+// When making authenticated preview requests, we go through the v4 Workers API rather than
+// hitting the preview service directly, so its response is formatted like a v4 API response.
+// These structs are here to convert from this format into the Preview defined above.
+#[derive(Debug, Deserialize)]
+struct ApiPreview {
+    pub preview_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+struct V4ApiResponse {
+    pub result: ApiPreview,
+}
+
 pub fn preview(
-    project: &Project,
-    method: Result<HTTPMethod, failure::Error>,
+    mut project: Project,
+    user: Option<GlobalUser>,
+    method: HTTPMethod,
     body: Option<String>,
 ) -> Result<(), failure::Error> {
-    let create_address = "https://cloudflareworkers.com/script";
+    let client: Client;
+
+    let preview = match &user {
+        Some(user) => {
+            log::info!("Running in Auth'd mode");
+
+            project.validate(false)?;
+
+            commands::build(&project)?;
+            client = http::auth_client(&user);
+
+            authenticated_upload(&client, &project)?
+        }
+        None => {
+            log::info!("Running in Un-auth'd mode");
+
+            // KV namespaces are not supported by the preview service unless you authenticate
+            // so we omit them and provide the user with a little guidance. We don't error out, though,
+            // because there are valid workarounds for this for testing purposes.
+            if project.kv_namespaces.is_some() {
+                message::warn("KV Namespaces are not supported in unauthenticated mode");
+                message::help(
+                    "Run wrangler config or set $CF_API_KEY and $CF_EMAIL to configure your user.",
+                );
+                project.kv_namespaces = None;
+            }
+
+            commands::build(&project)?;
+            client = http::client();
+
+            unauthenticated_upload(&client, &project)?
+        }
+    };
+
+    let worker_res = call_worker(&client, preview, method, body)?;
+
+    let msg = format!("Your worker responded with: {}", worker_res);
```

Comment:
Does this need a ✨ emoji?

Reply:
i mean idk, this is how it is today.

Reply:
oh actually it has the worker emojis.. i wonder if that gets added elsewhere

```
👷  Your worker responded with: Hello worker! 👷 
```